users/models.py

```diff
@@ -1,3 +1,290 @@
+
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name', 'user_phone']
+
+    def __str__(self):
+        """Return a string representation of the user."""
+        return self.email
+
+    def get_full_name(self):
+        """Return the user's full name."""
+        first_name = self.first_name.strip() or 'Unknown first name'
+        last_name = self.last_name.strip() or 'Unknown last name'
+        return f"{first_name} {last_name}".strip()
+
+    def get_roles_display(self):
+        """Return a string representation of the user's roles."""
+        return ', '.join(role.name for role in self.roles.all())
+
+    def add_role(self, role_name):
+        """
+        Add a role to the user.
+
+        Args:
+            role_name (str): Name of the role to add.
+
+        Raises:
+            ValidationError: If the role is invalid or already assigned.
+        """
+        if role_name not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Cannot add role: {role_name}")
+        if not self.is_active:
+            raise ValidationError("This account is not active.")
+        if self.is_soft_deleted:
+            raise ValidationError("This account is deleted.")
+        try:
+            role = Role.objects.get(name=role_name)
+        except Role.DoesNotExist:
+            raise ValidationError(f"Role '{role_name}' does not exist.")
+        if self.roles.filter(name=role_name).exists():
+            raise ValidationError(f"You already have the role: {role_name}")
+        self.roles.add(role)
+        self.save()
+        logger.info(f"Role '{role_name}' added to user {self.email}")
+
+    def remove_role(self, role_name):
+        """
+        Remove a role from the user.
+
+        Args:
+            role_name (str): Name of the role to remove.
+
+        Raises:
+            ValidationError: If the role is invalid or not assigned.
+        """
+        if role_name not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Cannot remove role: {role_name}")
+        if not self.is_active:
+            raise ValidationError("This account is not active.")
+        if self.is_soft_deleted:
+            raise ValidationError("This account is deleted.")
+        try:
+            role = self.roles.get(name=role_name)
+        except Role.DoesNotExist:
+            raise ValidationError(f"User does not have the role: {role_name}")
+        self.roles.remove(role)
+        self.save()
+        logger.info(f"Role '{role_name}' removed from user {self.email}")
+
+    def has_role(self, role_name):
+        """
+        Check if the user has a specific role.
+
+        Args:
+            role_name (str): Name of the role to check.
+
+        Returns:
+            bool: True if the user has the role, False otherwise.
+        """
+        if role_name not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Role '{role_name}' is not allowed.")
+        return self.roles.filter(name=role_name).exists()
```

Comment:
If you need to frequently check for roles like "Admin", "Investor", etc., you could cache these checks to avoid repeatedly querying the database.
_Example_
```
def has_role(self, role_name):
    if hasattr(self, '_cached_roles'):
        return role_name in self._cached_roles
    roles = self.roles.values_list('name', flat=True)
    self._cached_roles = list(roles)
    return role_name in self._cached_roles
```